src/main/scala/io/renku/jsonld/Cursor.scala

```diff
@@ -111,87 +120,167 @@ abstract class Cursor {
-    override lazy val top:    Option[JsonLD] = parent.top
+  private[jsonld] final case class PropertyCursor(parent: Cursor, property: Property, jsonLD: JsonLD)(implicit
+      decodingCache:                                      DecodingCache
+  ) extends Cursor {
+    override lazy val delete:   Cursor         = DeletedPropertyCursor(parent, property)
+    override lazy val top:      Option[JsonLD] = parent.top
+    override lazy val focusTop: Cursor         = top.map(TopCursor(_)).getOrElse(this)
   }
 
-  private[jsonld] final case class ListItemCursor(parent: Cursor, jsonLD: JsonLD) extends Cursor {
-    override lazy val top:    Option[JsonLD] = parent.top
-    override lazy val delete: Cursor         = Empty()
+  private[jsonld] final case class ListItemCursor(parent: Cursor, jsonLD: JsonLD)(implicit decodingCache: DecodingCache)
+      extends Cursor {
+    override lazy val delete:   Cursor         = Empty()
+    override lazy val top:      Option[JsonLD] = parent.top
+    override lazy val focusTop: Cursor         = top.map(TopCursor(_)).getOrElse(this)
   }
 
-  private[jsonld] final case class ArrayCursor(parent: Cursor, jsonLD: JsonLDArray) extends Cursor {
-    override lazy val top:    Option[JsonLD] = parent.top
-    override lazy val delete: Cursor         = Empty()
+  private[jsonld] object ListItemCursor {
+    def from(parent: Cursor, jsonLD: JsonLD): ListItemCursor = ListItemCursor(parent, jsonLD)(parent.decodingCache)
+  }
+
+  private[jsonld] final case class ArrayCursor(parent: Cursor, jsonLD: JsonLDArray)(implicit
+      decodingCache:                                   DecodingCache
+  ) extends Cursor {
+    override lazy val delete:   Cursor         = Empty()
+    override lazy val top:      Option[JsonLD] = parent.top
+    override lazy val focusTop: Cursor         = top.map(TopCursor(_)).getOrElse(this)
   }
 
   private[jsonld] final case class FlattenedArrayCursor(parent:      Cursor,
                                                         jsonLD:      JsonLDArray,
                                                         allEntities: Map[EntityId, JsonLDEntity]
-  ) extends Cursor {
-    override lazy val top:    Option[JsonLD] = parent.top
+  )(implicit decodingCache:                                          DecodingCache)
+      extends Cursor {
     override lazy val delete: Cursor         = Empty()
+    override lazy val top:    Option[JsonLD] = parent.top
+    override lazy val focusTop: Cursor = top
+      .map {
+        case jsonLD: JsonLDArray => FlattenedArrayCursor(this, jsonLD, allEntities)
+        case jsonLD => TopCursor(jsonLD)
+      }
+      .getOrElse(this)
+
+    def downTo(jsonLD: JsonLD): FlattenedJsonCursor = FlattenedJsonCursor(this, jsonLD, allEntities)
+  }
+}
+
+private[jsonld] sealed abstract class Caching(private[jsonld] val decodingCache: DecodingCache) {
+  self: Cursor =>
+
+  def findInCache[A](decoder: JsonLDDecoder[A]): Option[A] =
```

Comment:
Perhaps it will become clear after the code reivew, but what's the use case of a `findInCache` method without an entityId?